File: packages/drop-ai/android/src/main/java/com/drop/ai/NativeEmbed.kt
package com.drop.ai

import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Native mirror of `demoEmbedText` in src/convex/ai/demo.ts.
 *
 * The exact same deterministic 128-dim FNV-1a character n-gram algorithm, so
 * vectors computed on-device are directly comparable (cosine) with vectors
 * computed by the server — semantic search works with zero configuration,
 * zero keys and zero downloaded models.
 */
object NativeEmbed {
    private const val DIM = 128
    private val CLEAN_RE = Regex("[^a-z0-9\\s\u20ac\$\u00a3\u00a5]")
    private val SPACE_RE = Regex("\\s+")

    fun embed(text: String): FloatArray {
        val vec = FloatArray(DIM)
        val clean = CLEAN_RE.replace(text.lowercase(), " ")
        val tokens = SPACE_RE.split(clean).filter { it.isNotEmpty() }

        fun add(key: String) {
            var h = 2166136261L
            for (i in key.indices) {
                h = (h xor key[i].code.toLong()) and 0xFFFFFFFFL
                h = (h * 16777619L) and 0xFFFFFFFFL
            }
            // h.toInt() reproduces JS 32-bit signed semantics of Math.imul.
            val idx = abs(h.toInt()) % DIM
            vec[idx] += 1f
        }

        for (t in tokens) {
            add(t)
            if (t.length > 2) add("2:" + t.substring(0, 2))
            if (t.length > 3) add("3:" + t.substring(0, 3))
        }

        var mag = 0.0
        for (v in vec) mag += v * v
        mag = sqrt(mag)
        if (mag < 1e-9) return vec
        for (i in vec.indices) vec[i] = (vec[i] / mag).toFloat()
        return vec
    }
}

File: packages/drop-ai/ios/Sources/DropAI/NativeEmbed.swift
import Foundation

/// Native mirror of `demoEmbedText` in src/convex/ai/demo.ts.
///
/// The same deterministic 128-dim FNV-1a character n-gram algorithm used by
/// the server, so on-device vectors are cosine-comparable with server-side
/// vectors — semantic search works with zero configuration.
public enum NativeEmbed {
    private static let dim = 128
    private static let invalidChars = CharacterSet.alphanumerics
        .union(CharacterSet(charactersIn: " \u{20ac}$\u{00a3}\u{00a5}"))
        .inverted

    public static func embed(_ text: String) -> [Float] {
        var vec = [Float](repeating: 0, count: dim)
        let clean = text.lowercased()
            .components(separatedBy: invalidChars)
            .joined(separator: " ")
        let tokens = clean.split(whereSeparator: \.isWhitespace).map(String.init)

        func add(_ key: String) {
            var h: UInt64 = 2_166_136_261
            for scalar in key.utf16 {
                h ^= UInt64(scalar)
                h &*= 16_777_619
            }
            // Reproduce JS 32-bit signed semantics: take low 32 bits as signed.
            let signed = Int32(truncatingIfNeeded: UInt32(truncatingIfNeeded: h))
            let idx = abs(signed) % Int32(dim)
            vec[Int(idx)] += 1
        }

        for t in tokens {
            add(t)
            if t.count > 2 { add("2:" + String(t.prefix(2))) }
            if t.count > 3 { add("3:" + String(t.prefix(3))) }
        }

        let mag = sqrt(vec.reduce(0) { $0 + $1 * $1 })
        guard mag > 1e-9 else { return vec }
        return vec.map { $0 / mag }
    }
}
